VariantEncoding.md

```diff
@@ -0,0 +1,429 @@
+          |                       |
+          +-----------------------+
+          |                       |
+          :        offset         :  <-- little-endian, `offset_size` bytes
+          |                       |
+          +-----------------------+
+                      :
+          +-----------------------+
+          |                       |
+          :        offset         :  <-- little-endian, `offset_size` bytes
+          |                       |      (`dictionary_size + 1` offsets)
+          +-----------------------+
+          |                       |
+          :         bytes         :
+          |                       |
+          +-----------------------+
+```
+
+The metadata is encoded first with the `header` byte, then `dictionary_size` which is a little-endian value of `offset_size` bytes, and represents the number of string values in the dictionary.
+Next, is an `offset` list, which contains `dictionary_size + 1` values.
+Each `offset` is a little-endian value of `offset_size` bytes, and represents the starting byte offset of the i-th string in `bytes`.
+The first `offset` value will always be `0`, and the last `offset` value will always be the total length of `bytes`.
+The last part of the metadata is `bytes`, which stores all the string values in the dictionary.
+
+## Metadata encoding grammar
+
+The grammar for encoded metadata is as follows
+
+```
+metadata: <header> <dictionary_size> <dictionary>
+header: 1 byte (<version> | <sorted_strings> << 4 | (<offset_size_minus_one> << 6))
+version: a 4-bit version ID. Currently, must always contain the value 1
+sorted_strings: a 1-bit value indicating whether metadata strings are sorted
+offset_size_minus_one: 2-bit value providing the number of bytes per dictionary size and offset field.
+dictionary_size: `offset_size` bytes. little-endian value indicating the number of strings in the dictionary
+dictionary: <offset>* <bytes>
+offset: `offset_size` bytes. little-endian value indicating the starting position of the ith string in `bytes`. The list should contain `dictionary_size + 1` values, where the last value is the total length of `bytes`.
+bytes: dictionary string values
+```
+
+Notes:
+- Offsets are relative to the start of the `bytes` array.
+- The length of the ith string can be computed as `offset[i+1] - offset[i]`.
+- The offset of the first string is always equal to 0 and is therefore redundant. It is included in the spec to simplify in-memory-processing.
+- `offset_size_minus_one` indicates the number of bytes per `dictionary_size` and `offset` entry. I.e. a value of 0 indicates 1-byte offsets, 1 indicates 2-byte offsets, 2 indicates 3 byte offsets and 3 indicates 4-byte offsets.
+- If `sorted_strings` is set to 1, strings in the dictionary must be unique and sorted in lexicographic order. If the value is set to 0, readers may not make any assumptions about string order or uniqueness.
```

Comment:
does this assume any kind of encoding or is it byte-wise?

Reply:
All strings are UTF-8, but I think it's a follow up to clarify that.

Reply:
Yes, they are all UTF-8. I can add a follow up to clarify that point.

Reply:
so lexicographic order is defined by unicode code points.